src/groth16/verifier/batch.rs

```diff
@@ -167,4 +173,117 @@ where
             Err(VerificationError::InvalidProof)
         }
     }
+
+    /// Perform batch verification with a particular `VerifyingKey`, returning
+    /// `Ok(())` if all proofs were verified and `VerificationError` otherwise.
+    ///
+    /// This performs the bulk of internal arithmetic over the global rayon
+    /// threadpool.
+    #[cfg(feature = "multicore")]
+    #[allow(non_snake_case)]
+    pub fn verify_multicore(self, vk: &VerifyingKey<E>) -> Result<(), VerificationError> {
+        if self
+            .items
+            .iter()
+            .any(|Item { inputs, .. }| inputs.len() + 1 != vk.ic.len())
+        {
+            return Err(VerificationError::InvalidVerifyingKey);
+        }
+
+        struct Accumulator<E: MultiMillerLoop> {
+            gammas: Vec<E::Fr>,
+            delta: E::G1,
+            y: E::Fr,
+            ml_result: Option<E::Result>,
+        }
+
+        impl<E: MultiMillerLoop> Accumulator<E> {
+            fn new(ic_len: usize) -> Self {
+                Accumulator {
+                    gammas: vec![E::Fr::zero(); ic_len],
+                    delta: E::G1::identity(),
+                    y: E::Fr::zero(),
+                    ml_result: None,
+                }
+            }
+        }
+
+        let ic_len = vk.ic.len();
+
+        let acc = self
+            .items
+            // This chunk size was obtained heuristically.
+            .par_chunks(8)
+            .map(|items| {
+                let mut acc = Accumulator::<E>::new(ic_len);
+                let mut ml_terms: Vec<(E::G1Affine, E::G2Prepared)> = vec![];
+                let z = loop {
+                    let z = E::Fr::random(&mut OsRng);
+                    if !z.is_zero_vartime() {
+                        break z;
+                    }
+                };
```

Comment:
Btw, the spec explicitly says that you're not required to exclude $z = 0$:
![image](https://user-images.githubusercontent.com/643204/177219426-bc3d7d08-69e5-4536-a0cd-c81a81fc5eb2.png)
```suggestion
                let z = E::Fr::random(&mut OsRng);
```
The same simplification could be made on lines 126-131.

(The probability of any of the $z_j$ being $0$, over $N$ proofs, is bounded above by $N / 2^{128}$. This will be larger than $2^{-125}$, but the $125$-bit security target refers to the adversary's work factor, not its success probability.)

Reply:
I don't believe the spec had that non-normative note at the time the original batch validation logic was written (per the comment in that code).

> I don't know whether you want to improve performance further by sampling 16 random bytes and converting to E::Fr?

This won't improve performance, because we use constant time scalar multiplication:
- https://github.com/zkcrypto/bls12_381/blob/34dab74ae44cd77dfa867928dd2e6c8383d8dc9d/src/g1.rs#L737-L751
- https://github.com/zkcrypto/bls12_381/blob/34dab74ae44cd77dfa867928dd2e6c8383d8dc9d/src/g2.rs#L808-L822

Reply:
> I don't believe the spec had that non-normative note at the time the original batch validation logic was written (per the comment in that code).

Right, it was added in v2021.2.17 (https://github.com/zcash/zips/commit/4da403f470c1708c807938359c284022fcb48f6f).